File: src/main/java/com/mbayou/kick4k/KickClient.java
package com.mbayou.kick4k;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.sun.net.httpserver.HttpServer;
import com.mbayou.kick4k.api.ApiResponse;
import com.mbayou.kick4k.api.ApiResponseDeserializer;
import com.mbayou.kick4k.authorization.AuthorizationClient;
import com.mbayou.kick4k.authorization.RefreshTokenStore;
import com.mbayou.kick4k.categories.CategoriesClient;
import com.mbayou.kick4k.channels.ChannelsClient;
import com.mbayou.kick4k.chat.ChatClient;
import com.mbayou.kick4k.events.EventsClient;
import com.mbayou.kick4k.events.handler.BuiltInKickWebhookHandler;
import com.mbayou.kick4k.events.handler.KickEventDispatcher;
import com.mbayou.kick4k.events.handler.KickSignatureVerifier;
import com.mbayou.kick4k.livestreams.LivestreamsClient;
import com.mbayou.kick4k.moderation.ModerationClient;
import com.mbayou.kick4k.publicKey.PublicKeyClient;
import com.mbayou.kick4k.users.UsersClient;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.http.HttpClient;

public class KickClient {
    private final AuthorizationClient authorizationClient;
    private final CategoriesClient categoriesClient;
    private final UsersClient usersClient;
    private final ChannelsClient channelsClient;
    private final ChatClient chatClient;
    private final ModerationClient moderationClient;
    private final LivestreamsClient livestreamsClient;
    private final PublicKeyClient publicKeyClient;
    private final EventsClient eventsClient;
    private final KickSignatureVerifier signatureVerifier;
    private final KickEventDispatcher eventDispatcher;

    private HttpServer httpServer;

    public KickClient(KickConfiguration configuration) {
        HttpClient httpClient = HttpClient.newHttpClient();

        SimpleModule serializerModule = new SimpleModule()
                .addDeserializer(ApiResponse.class, new ApiResponseDeserializer<>());
        ObjectMapper objectMapper = new ObjectMapper()
                .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
                .registerModule(new JavaTimeModule())
                .registerModule(serializerModule);

        this.authorizationClient = new AuthorizationClient(httpClient, objectMapper, configuration, configuration.getTokenStore());
        this.categoriesClient = new CategoriesClient(httpClient, objectMapper, configuration, this.authorizationClient);
        this.usersClient = new UsersClient(httpClient, objectMapper, configuration, this.authorizationClient);
        this.channelsClient = new ChannelsClient(httpClient, objectMapper, configuration, this.authorizationClient);
        this.chatClient = new ChatClient(httpClient, objectMapper, configuration, this.authorizationClient);
        this.moderationClient = new ModerationClient(httpClient, objectMapper, configuration, this.authorizationClient);
        this.livestreamsClient = new LivestreamsClient(httpClient, objectMapper, configuration, this.authorizationClient);
        this.publicKeyClient = new PublicKeyClient(httpClient, objectMapper, configuration, this.authorizationClient);
        this.eventsClient = new EventsClient(httpClient, objectMapper, configuration, this.authorizationClient);

        this.signatureVerifier = new KickSignatureVerifier(this.publicKeyClient);
        this.eventDispatcher = new KickEventDispatcher(objectMapper);
    }

    public void startWebhookReceiver(String path, int port) {
        try {
            this.httpServer = HttpServer.create(new InetSocketAddress(port), 0);
            this.httpServer.createContext(path, new BuiltInKickWebhookHandler(this.signatureVerifier, this.eventDispatcher));
            this.httpServer.setExecutor(null);
            this.httpServer.start();
        } catch (IOException exception) {
            throw new RuntimeException("Failed to start web server", exception);
        }
    }

    public void stopWebhookReceiver() {
        if (this.httpServer == null) {
            throw new IllegalStateException("Web server is not running");
        }

        this.httpServer.stop(100);
    }

    public AuthorizationClient authorization() {
        return authorizationClient;
    }

    public CategoriesClient categories() {
        return categoriesClient;
    }

    public UsersClient users() {
        return usersClient;
    }

    public ChannelsClient channels() {
        return channelsClient;
    }

    public ChatClient chat() {
        return chatClient;
    }

    public ModerationClient moderation() {
        return moderationClient;
    }

    public LivestreamsClient livestreams() {
        return livestreamsClient;
    }

    public PublicKeyClient publicKey() {
        return publicKeyClient;
    }

    public EventsClient events() {
        return eventsClient;
    }

    public KickSignatureVerifier signatureVerifier() {
        return signatureVerifier;
    }

    public KickEventDispatcher eventDispatcher() {
        return eventDispatcher;
    }
}

File: src/main/kotlin/com/mbayou/kick4k/KickClient.kt
package com.mbayou.kick4k

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.databind.module.SimpleModule
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.sun.net.httpserver.HttpServer
import com.mbayou.kick4k.api.ApiResponse
import com.mbayou.kick4k.api.ApiResponseDeserializer
import com.mbayou.kick4k.authorization.AuthorizationClient
import com.mbayou.kick4k.categories.CategoriesClient
import com.mbayou.kick4k.channels.ChannelsClient
import com.mbayou.kick4k.chat.ChatClient
import com.mbayou.kick4k.events.EventsClient
import com.mbayou.kick4k.events.handler.BuiltInKickWebhookHandler
import com.mbayou.kick4k.events.handler.KickEventDispatcher
import com.mbayou.kick4k.events.handler.KickSignatureVerifier
import com.mbayou.kick4k.livestreams.LivestreamsClient
import com.mbayou.kick4k.moderation.ModerationClient
import com.mbayou.kick4k.publicKey.PublicKeyClient
import com.mbayou.kick4k.users.UsersClient
import java.io.IOException
import java.net.InetSocketAddress
import java.net.http.HttpClient

class KickClient(configuration: KickConfiguration) {
    private val authorizationClient: AuthorizationClient
    private val categoriesClient: CategoriesClient
    private val usersClient: UsersClient
    private val channelsClient: ChannelsClient
    private val chatClient: ChatClient
    private val moderationClient: ModerationClient
    private val livestreamsClient: LivestreamsClient
    private val publicKeyClient: PublicKeyClient
    private val eventsClient: EventsClient
    private val signatureVerifier: KickSignatureVerifier
    private val eventDispatcher: KickEventDispatcher

    private var httpServer: HttpServer? = null

    init {
        val httpClient = HttpClient.newHttpClient()

        val serializerModule = SimpleModule()
            .addDeserializer(ApiResponse::class.java, ApiResponseDeserializer<Any>())

        val objectMapper = ObjectMapper()
            .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
            .registerModule(JavaTimeModule())
            .registerModule(serializerModule)

        this.authorizationClient = AuthorizationClient(httpClient, objectMapper, configuration)
        this.categoriesClient = CategoriesClient(httpClient, objectMapper, configuration)
        this.usersClient = UsersClient(httpClient, objectMapper, configuration)
        this.channelsClient = ChannelsClient(httpClient, objectMapper, configuration)
        this.chatClient = ChatClient(httpClient, objectMapper, configuration)
        this.moderationClient = ModerationClient(httpClient, objectMapper, configuration)
        this.livestreamsClient = LivestreamsClient(httpClient, objectMapper, configuration)
        this.publicKeyClient = PublicKeyClient(httpClient, objectMapper, configuration)
        this.eventsClient = EventsClient(httpClient, objectMapper, configuration)

        this.signatureVerifier = KickSignatureVerifier(publicKeyClient)
        this.eventDispatcher = KickEventDispatcher(objectMapper)
    }

    fun startWebhookReceiver(path: String, port: Int) {
        try {
            httpServer = HttpServer.create(InetSocketAddress(port), 0).apply {
                createContext(path, BuiltInKickWebhookHandler(signatureVerifier, eventDispatcher))
                executor = null
                start()
            }
        } catch (exception: IOException) {
            throw RuntimeException("Failed to start web server", exception)
        }
    }

    fun stopWebhookReceiver() {
        val server = httpServer ?: throw IllegalStateException("Web server is not running")
        server.stop(100)
        httpServer = null
    }

    fun authorization(): AuthorizationClient = authorizationClient
    fun categories(): CategoriesClient = categoriesClient
    fun users(): UsersClient = usersClient
    fun channels(): ChannelsClient = channelsClient
    fun chat(): ChatClient = chatClient
    fun moderation(): ModerationClient = moderationClient
    fun livestreams(): LivestreamsClient = livestreamsClient
    fun publicKey(): PublicKeyClient = publicKeyClient
    fun events(): EventsClient = eventsClient
    fun signatureVerifier(): KickSignatureVerifier = signatureVerifier
    fun eventDispatcher(): KickEventDispatcher = eventDispatcher
}
